### PR TITLE
kernel: generate placeholders for kobj tables before final build

### PR DESCRIPTION
**Pull request by dcpleung**
Due to the use of gperf to generate hash table for kobjects,
the addresses of these kobjects cannot change during the last
few phases of linking (especially between zephyr_prebuilt.elf
and zephyr.elf). Because of this, the gperf generated data
needs to be placed at the end of memory to avoid pushing symbols
around in memory. This prevents moving these generated blocks
to earlier sections, for example, pinned data section needed
for demand paging. So create placeholders for use in
intermediate linking to reserve space for these generated blocks.
Due to uncertainty on the size of these blocks, more space is
being reserved which could result in wasted space. Though, this
retains the use of hash table for faster lookup.

Alternative to #33719

### CODE REVIEW

**Comment by dcpleung** [comment]
Related to #32129

**Comment by pabigot** [comment]
This and #33719 seem related to the problem underlying https://github.com/zephyrproject-rtos/zephyr/issues/32129 ; it may be worth looking at the description there when considering alternative solutions, as existing kernel objects are not the only data structures that might change size between link phases.

**Comment by dcpleung** [comment]
Updated to use some type of fudging factor as proposed by @mbolivar-nordic. The multiple build passes does not guarantee enough space is reserved for final build so this new method is better and this won't lengthen the build (that much).

**Comment by mbolivar-nordic** [comment]
The idea seems sound to me, though I suspect if a fudge factor of 200% is really required in practice, we may want to drop gperf eventually for footprint reasons.

File: arch/Kconfig [hunk 1]
```diff
@@ -281,6 +281,28 @@ config KOBJECT_TEXT_AREA
 	help
 	  Size of kernel object text area. Used in linker script.
 
+config KOBJECT_DATA_AREA_RESERVE_MULTIPLIER
+	int "Multiplier for reserving kobject data area"
+	default 200
```
Comment (nashif) [inline arch/Kconfig:286]: shouldn't this be 2 instead? or you really want to say 200 times larger?

Comment (dcpleung) [reply arch/Kconfig:286]: 200%. So it is possible to specify 120% or 150%. Kconfig doesn't take numbers with decimals.

**Comment by dcpleung** [comment]
> The idea seems sound to me, though I suspect if a fudge factor of 200% is really required in practice, we may want to drop gperf eventually for footprint reasons.

200% is simply a safe choice for most scenarios. Apps can tune the value if needed.

**Comment by dcpleung** [comment]
Appreciate some reviews to get this or #33719 merged.

File: arch/Kconfig (hunk 1 above)
Comment (pfalcon) [reply arch/Kconfig:286]: Then apparently it should say "percent" somewhere in the description. And maybe even in the option name, and then maybe use standard percentage's additive semantics, where [+]0% means "zero extra overhead" and [+]100% means "2x times the original size".

Comment (dcpleung) [reply arch/Kconfig]: Updated.

**Review by nashif — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**